sys/arm64/arm64/pmap.c

```diff
@@ -4798,6 +4854,321 @@ pmap_sync_capdirty(pmap_t pmap)
+{
+	struct rwlock *lock;
+	pv_entry_t pv;
+	struct md_page *pvh;
+	pt_entry_t pte;
+	pmap_t pmap;
+	int lvl, md_gen, pvh_gen;
+	boolean_t rv;
+
+	rv = TRUE;
+	lock = VM_PAGE_TO_PV_LIST_LOCK(m);
+	rw_rlock(lock);
+restart:
+	TAILQ_FOREACH(pv, &m->md.pv_list, pv_next) {
+		pmap = PV_PMAP(pv);
+		PMAP_ASSERT_STAGE1(pmap);
+		if (!PMAP_TRYLOCK(pmap)) {
+			md_gen = m->md.pv_gen;
+			rw_runlock(lock);
+			PMAP_LOCK(pmap);
+			rw_rlock(lock);
+			if (md_gen != m->md.pv_gen) {
+				PMAP_UNLOCK(pmap);
+				goto restart;
+			}
+		}
+		pte = pmap_load(pmap_pte(pmap, pv->pv_va, &lvl));
+		PMAP_UNLOCK(pmap);
+		KASSERT(lvl == 3,
+		    ("pmap_page_test_mappings: Invalid level %d", lvl));
+		if (pte & ATTR_CDBM) {
+			rv = FALSE;
+			goto out;
+		}
+		KASSERT((pte & ATTR_SC) == 0, ("ATTR_SC without ATTR_CDBM"));
+	}
+	if ((m->flags & PG_FICTITIOUS) == 0) {
+		pvh = page_to_pvh(m);
+		TAILQ_FOREACH(pv, &pvh->pv_list, pv_next) {
+			pmap = PV_PMAP(pv);
+			PMAP_ASSERT_STAGE1(pmap);
+			if (!PMAP_TRYLOCK(pmap)) {
+				md_gen = m->md.pv_gen;
+				pvh_gen = pvh->pv_gen;
+				rw_runlock(lock);
+				PMAP_LOCK(pmap);
+				rw_rlock(lock);
+				if (md_gen != m->md.pv_gen ||
+				    pvh_gen != pvh->pv_gen) {
+					PMAP_UNLOCK(pmap);
+					goto restart;
+				}
+			}
+			pte = pmap_load(pmap_pte(pmap, pv->pv_va, &lvl));
+			PMAP_UNLOCK(pmap);
+			KASSERT(lvl == 2,
+			    ("pmap_page_test_mappings: Invalid level %d", lvl));
+			if (pte & ATTR_CDBM) {
+				rv = FALSE;
+				goto out;
+			}
+			KASSERT((pte & ATTR_SC) == 0,
+			    ("ATTR_SC without ATTR_CDBM"));
+		}
+	}
+out:
+	rw_runlock(lock);
+	/*
+	 * It's important that we test the PGA_CAPDIRTY flag again *after*
+	 * we've looked at all the PTEs: we might have raced a removal of a PTE
+	 * that had ATTR_CDBM set.
+	 *
+	 * Despite not holding a pmap LOCKed right now, something is preventing
+	 * new mappings (PMAP_CAPLOADGEN_EXCLUSIVE) and so we don't need to
+	 * worry that we're missing something here.
+	 */
+	if (rv && !(vm_page_astate_load(m).flags & PGA_CAPDIRTY)) {
+		vm_page_aflag_clear(m, PGA_CAPSTORE);
+	}
+}
+
+int
+pmap_caploadgen_update(pmap_t pmap, vm_offset_t *pva, vm_page_t *mp, int flags)
+{
+	enum pmap_caploadgen_res res;
+	pt_entry_t *pte, tpte, exppte;
+	vm_page_t m;
+	vm_offset_t next = 0, va = *pva;
+	int lvl;
+
+	PMAP_ASSERT_STAGE1(pmap);
+	PMAP_LOCK(pmap);
+
+	KASSERT(!(READ_SPECIALREG(cctlr_el1) & CCTLR_EL1_TGEN0_MASK) ==
+	    !(pmap->flags.uclg),
+	    ("pmap_caploadgen_update: pmap clg %d but CPU mismatch",
+	    (int)pmap->flags.uclg));
+
+	pte = pmap_pte(pmap, va, &lvl);
+	if (pte == NULL) {
+		m = NULL;
+		res = PMAP_CAPLOADGEN_UNABLE;
+		goto out;
+	}
+	tpte = pmap_load(pte);
+
+	switch (tpte & ATTR_LC_MASK) {
+	case ATTR_LC_DISABLED:	// tag clearing
+	case ATTR_LC_ENABLED:	// always allowed; not for us to worry about?
+		m = NULL;
+		res = PMAP_CAPLOADGEN_UNABLE;
+		goto out;
+	case ATTR_LC_GEN0:	// ah, here we go
+	case ATTR_LC_GEN1:
+		break;
+	}
+
+	switch(lvl) {
```

Comment:
On the RISC-V side, we found it more or less necessary to always break superpages apart; that is, `pmap_caploadgen_update` bails if the page radix isn't the smallest (4K).  What can happen is that the VM will assemble superpages that span multiple `vm_object`s, and so the simplistic iteration in `vm_caprevoke.c` would become very alarmed when its next address was past the end of the object it was considering.  Moreover, the `PGA_*` bits for superpages are diffused between the `vm_page`s that were composited into the superpage, so `pmap_caploadgen_test_all_clean` isn't doing the right thing as it stands (which is entirely my fault, you just had the misfortune of copying my bugs).

That's not exactly a stellar state of affairs, but it's where the code is right now.  The VM doesn't really have, AFAIK, any MI notion that superpages exist, which makes doing much better somewhat tricky (though I am quite open to suggestions).

Reply:
Ah, this is what I get for reviewing in commit order when you have so faithfully copied commits over!